dll/ntdll/ldr/ldrinit.c

```diff
@@ -1538,7 +1538,36 @@ VOID
 NTAPI
 LdrpValidateImageForMp(IN PLDR_DATA_TABLE_ENTRY LdrDataTableEntry)
 {
-    UNIMPLEMENTED;
+    PIMAGE_NT_HEADERS NtHeaders;
+    NTSTATUS Status;
+
+    /* Get the NT Headers */
+    NtHeaders = RtlImageNtHeader(LdrDataTableEntry->DllBase);
+    if (NtHeaders == NULL)
+    {
+        RtlRaiseStatus(STATUS_INVALID_IMAGE_FORMAT);
+    }
+
+    /* Check if the image is UP only */
+    if (NtHeaders->FileHeader.Characteristics & IMAGE_FILE_UP_SYSTEM_ONLY)
+    {
+        /* Set process affinity to 1 */
+        DPRINT1("LDR: Image is UP only, setting process affinity to 1\n");
+        KAFFINITY AffinityMask = 1;
+        Status = NtSetInformationProcess(NtCurrentProcess(),
+                                         ProcessAffinityMask,
+                                         &AffinityMask,
+                                         sizeof(AffinityMask));
```

Comment:
This means an UP image will run specifically on the first processor / thread?

Reply:
That's the idea, yes.

Reply:
That's no longer the idea, because the kernel does this already.